### DISCUSSION TITLE
Predict Function Issues

### DISCUSSION THREAD

**Discussion by Psychodogs** [pdGruby/geokrige#3] [Q&A] [answered]
Was using the geokrige module for some spatial geometry interpolation, and tried to use the geopandas package to read a shapefile to a msehgrid.
```
import geopandas as gpd
from geokrige.tools import TransformerGDF

shp_file = 'shapefile/qj614bt0216.shp'
prediction_gdf = gpd.read_file(shp_file).to_crs(crs='EPSG:4326')

transformer = TransformerGDF()
transformer.load(prediction_gdf)

meshgrid = transformer.meshgrid(density=2)
mask = transformer.mask()`
```
This was what I used, according to the documentation page. But this error popped up:
```
raise TypeError(f"The type of the 'X' argument is invalid! Valid types are [numpy.ndarray, list], "
TypeError: The type of the 'X' argument is invalid! Valid types are [numpy.ndarray, list], got <class 'tuple'> instead
```
I assumed it was some sort of issue with the shapefile, and tried troubleshooting with no success, so I tried the tutorial code: 
import numpy as np
```

m = np.array([14.5, 15.5, 17.5, 20.5, 23.2, 22])
n = np.array([50.2, 51.7, 52.3, 54.5, 50.5, 53])

X = np.column_stack([m, n])
y = np.array([5, 10, 11, 7, 6.5, 9])

from geokrige.methods import SimpleKriging

kgn = SimpleKriging()
kgn.load(X, y)

kgn.variogram(bins=6, plot=False)
kgn.fit(plot=False)

import matplotlib.pyplot as plt

# Perform prediction on a mesh grid
lon = np.linspace(13.5, 24.5, 10)
lat = np.linspace(49.5, 55.5, 10)
meshgrid = np.meshgrid(lon, lat)

X, Y = meshgrid
Z = kgn.predict(meshgrid)

# Plot results
plt.scatter(m, n, color='red', zorder=5)
plt.pcolormesh(X, Y, Z, vmin=5, vmax=11)
plt.colorbar()

for i, value in enumerate(y):
    x_pos = m[i]
    y_pos = n[i]
    plt.text(x_pos, y_pos + 0.15, value, ha='center', fontweight='bold')

plt.title("GeoKrige - simple prediction result")
plt.show()
```
But the same error still pops up. I think the issue lies with how the numpy package handles the [numpy.meshgrid ](https://numpy.org/devdocs/reference/generated/numpy.meshgrid.html) function. The current version (1.7.0) of numpy outputs a tuple, which clashes with the input for the predict function, I am not sure how to change my output to fix the code.
### Edit:
I seemed to have fixed it by changing the source code, around line 387, where it was originally
```
if isinstance(X, list):  # np.meshgrid
            shape = tuple(X[0].shape)
to
if isinstance(X, tuple):  # np.meshgrid
            shape = tuple(X[0].shape)
```
and it seemed to work as the documentation showed.

**Comment by pdGruby** [comment, marked as answer]
Hello @Psychodogs,

Thank you so much for your valuable feedback! Your observations are greatly appreciated and will undoubtedly help improve the GeoKrige package. I’ve taken some time to investigate the issue, and here are my findings.

# Notes

> But the same error still pops up. I think the issue lies with how the numpy package handles the [numpy.meshgrid](https://numpy.org/devdocs/reference/generated/numpy.meshgrid.html) function.

You are absolutely correct. The issue arises because GeoKrige expects the `X` parameter in the `predict` method to be either a `list` or a `numpy.ndarray`. Historically, the `numpy.meshgrid` function returned **a list** of `ndarray` objects, but starting from version 2.0.0, it returns **a tuple** of `ndarray` objects. This change is what’s causing the error.

---

> The current version (1.7.0) of numpy outputs a tuple, which clashes with the input for the predict function, I am not sure how to change my output to fix the code.

Are you sure you’re using Numpy version 1.7.0? That version was released back in 2013, and according to the [documentation](https://docs.scipy.org/doc/numpy-1.7.0/reference/generated/numpy.meshgrid.html), `numpy.meshgrid` should still return a list of `ndarray` objects in version 1.7.0, which shouldn't trigger this error. **However, GeoKrige was developed with Numpy 1.24.3, and the `requirements.txt` specifies Numpy >= 1.24.3**.

---

> I seemed to have fixed it by changing the source code, around line 387, where it was originally

Your modification would likely make GeoKrige work with the latest version of Numpy, but I’d consider it more of a workaround for the moment than a proper fix. GeoKrige has been tested primarily with Numpy versions below 2.0.0, so making such changes could introduce instability in other areas of the package, as the differences between Numpy 1.x and 2.x seem to be significant.

If you'd like to continue using Numpy >=2.0.0, I suggest an alternative workaround: before passing the meshgrid to the `predict` method, convert it to a list using Python's `list` function. **However, I can’t guarantee that this will work seamlessly with other functionalities.**

# My Recommendation

### For now, I recommend using Numpy versions >= 1.24.3 and <= 1.26.4.

While the `requirements.txt` file specifies Numpy >=1.24.3, it's clear from your experience that this doesn’t hold for Numpy 2.x, which returns a tuple rather than a list of `ndarray` objects. Please, let me know if my assumption that you are using Numpy 2.x instead of Numpy 1.7.0 is correct!

# Next Steps

- The GeoKrige package should be updated to ensure compatibility with the latest versions of Numpy. While the changes you proposed could presumably be implemented, this issue definitely needs further investigation to assess the potential consequences..
- If immediate compatibility with Numpy 2.x is not feasible, the `requirements.txt` should be updated to reflect the supported versions.

Thanks again for bringing this topic to attention!

**Reply by Psychodogs** [reply]
I have just checked, and yes, I am using Numpy 2.0.1. I got the false value of 1.7.0 from the documentation page of numpy.
